From 791f38217c1bc27726509874e95f52a356427263 Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Sat, 15 Feb 2025 19:00:04 +0100
Subject: [PATCH 23/33] dashdec: ensure that pts always starts from zero

presentationTimeOffset can be used for the same way, but this is more generic solution
---
 libavformat/dashdec.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/libavformat/dashdec.c b/libavformat/dashdec.c
index d1e1612773..28eb383c0b 100644
--- a/libavformat/dashdec.c
+++ b/libavformat/dashdec.c
@@ -117,6 +117,7 @@ struct representation {
     uint32_t init_sec_data_len;
     uint32_t init_sec_buf_read_offset;
     int64_t cur_timestamp;
+    int64_t first_pts;
     int is_restart_needed;
 
     char *cenc_decryption_key;
@@ -892,6 +893,7 @@ static int parse_manifest_representation(AVFormatContext *s, const char *url,
 
     rep->parent = s;
     rep->type = type;
+    rep->first_pts = INT64_MIN;
     representation_segmenttemplate_node = find_child_node_by_name(representation_node, "SegmentTemplate");
     representation_baseurl_node = find_child_node_by_name(representation_node, "BaseURL");
     representation_segmentlist_node = find_child_node_by_name(representation_node, "SegmentList");
@@ -2346,6 +2348,19 @@ static int dash_read_packet(AVFormatContext *s, AVPacket *pkt)
     while (!ff_check_interrupt(c->interrupt_callback) && !ret) {
         ret = av_read_frame(cur->ctx, pkt);
         if (ret >= 0) {
+            if(cur->n_fragments != 1)
+            {
+                /* in multifragments mode - ensure, that PTS always starts from 0 */
+                if(cur->first_pts == INT64_MIN)
+                {
+                    av_log(c, AV_LOG_TRACE, "Setting first PTS to pos[%" PRId64 "] for %s\n", cur->first_pts, cur->id);
+                    cur->first_pts = pkt->pts;
+                }
+
+                pkt->pts -= cur->first_pts;
+                pkt->dts -= cur->first_pts;
+            }
+
             /* If we got a packet, return it */
             cur->cur_timestamp = av_rescale(pkt->pts, (int64_t)cur->ctx->streams[0]->time_base.num * 90000, cur->ctx->streams[0]->time_base.den);
             pkt->stream_index = cur->stream_index;
@@ -2398,6 +2413,8 @@ static int dash_seek(AVFormatContext *s, struct representation *pls, int64_t see
     // find the nearest fragment
     if (pls->n_timelines > 0 && pls->fragment_timescale > 0) {
         int64_t num = pls->first_seq_no;
+        seek_pos_msec += (pls->first_pts / pls->fragment_timescale) * 1000;
+
         av_log(pls->parent, AV_LOG_VERBOSE, "dash_seek with SegmentTimeline start n_timelines[%d] "
                "last_seq_no[%"PRId64"].\n",
                (int)pls->n_timelines, (int64_t)pls->last_seq_no);
-- 
2.45.2.windows.1

